app/gui/view/graph-editor/src/builtin/visualization/native/text_visualization/text_provider.rs

```diff
@@ -42,11 +45,460 @@ pub trait TextProvider {
 // === Text Provider FRP ===
 // ===========================
 
+type ColumnIndex = usize;
+/// Column Width in characters.
+type ColumnWidth = usize;
+/// Column Width in lines
+type RowHeight = usize;
+type RowIndex = usize;
+
+/// Specification of a table layout. Contains the column and row widths, as well as the column and
+/// row names. Provides methods for updating the specification and for accessing layout properties.
+/// We allow for a table specification to be partially specified, i.e., some of the information may
+/// be `None`. This is used to allow for lazy initialisation of the layout and incremental updates
+/// as the table is updated.
+#[derive(Clone, Debug, Deserialize, Serialize, Default)]
+pub struct TableSpecification {
+    /// The width of each column in characters.
```

Comment:
Please make subsection for TableSpecification, and the subsection for `define_endpoints_2` below.